File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/SwitchStyle.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;

public class SwitchStyle extends ToggleStyle {
    @NonNull
    private final Color onColor;
    @NonNull
    private final Color offColor;

    public SwitchStyle(
        @NonNull Color onColor,
        @NonNull Color offColor
    ) {
        super(ToggleType.SWITCH);

        this.onColor = onColor;
        this.offColor = offColor;
    }

    @NonNull
    public static SwitchStyle fromJson(@NonNull JsonMap json) throws JsonException {
        JsonMap colors = json.opt("toggle_colors").optMap();
        Color onColor = Color.fromJsonField(colors, "on");
        if (onColor == null) {
            throw new JsonException("Failed to parse SwitchStyle! Field 'toggle_colors.on' may not be null.");
        }
        Color offColor = Color.fromJsonField(colors, "off");
        if (offColor == null) {
            throw new JsonException("Failed to parse SwitchStyle! Field 'toggle_colors.off' may not be null.");
        }

        return new SwitchStyle(onColor, offColor);
    }

    @NonNull
    public Color getOnColor() {
        return onColor;
    }

    @NonNull
    public Color getOffColor() {
        return offColor;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/SwitchStyle.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import com.urbanairship.json.requireMap

public class SwitchStyle public constructor(
    @JvmField public val onColor: Color,
    @JvmField public val offColor: Color
) : ToggleStyle(ToggleType.SWITCH) {

    public companion object {

        private const val KEY_TOGGLE_COLORS = "toggle_colors"
        private const val KEY_ON = "on"
        private const val KEY_OFF = "off"

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): SwitchStyle {
            val content = json.requireMap()
            val colors = content.requireMap(KEY_TOGGLE_COLORS)

            return SwitchStyle(
                onColor = Color.fromJson(colors.require(KEY_ON)),
                offColor = Color.fromJson(colors.require(KEY_OFF))
            )
        }
    }
}
